QuantLib/ql/termstructures/yield/discountcurve.hpp

```diff
@@ -113,7 +114,12 @@ namespace QuantLib {
 
     template <class T>
     inline Date InterpolatedDiscountCurve<T>::maxDate() const {
-        return dates_.back();
+        // +7 is just a patch to partially (but safely) deal with a pillar date
+        // possibly not being equal to BootstrapHelper's maturity or latest
+        // relevant date. A better (non-backward-compatible) solution would be
+        //if (maxDate_ != Date())
+        //    return maxDate_;
```

Comment:
This might be written in a (very close to) backward-compatible way by adding `maxDate_` as a protected data member to the base `InterpolatedCurve` class. The bootstrapper can set it at the end, since it's a friend. A bit smelly (since the other data members are times) but I wouldn't be outraged.

The only way this could break is if someone had written a bootstrapped curve that doesn't inherit from `InterpolatedCurve`. But that's very unlikely, since its data members are required for bootstrap.